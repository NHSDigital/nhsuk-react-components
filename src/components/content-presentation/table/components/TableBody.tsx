import classNames from 'classnames';
import React, { FC, HTMLProps } from 'react';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const TableBody: FC<HTMLProps<HTMLTableSectionElement>> = ({ className, children, ...rest }) => (
  <tbody className={classNames('nhsuk-table__body', className)} {...rest}>
    <TableSectionContext.Provider value={TableSection.BODY}>
      {children}
    </TableSectionContext.Provider>
  </tbody>
);
TableBody.displayName = 'Table.Body';

export default TableBody;
