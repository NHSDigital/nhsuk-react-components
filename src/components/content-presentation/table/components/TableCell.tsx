import classNames from 'classnames';
import React, { FC, HTMLProps, useContext } from 'react';
import useDevWarning from '@util/hooks/UseDevWarning';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const CellOutsideOfSectionWarning =
  'Table.Cell used outside of a Table.Head or Table.Body component. Unable to determine section type from context.';

export interface TableCellProps extends HTMLProps<HTMLTableCellElement> {
  _responsive?: boolean;
  _responsiveHeading?: string;
  isNumeric?: boolean;
}

const TableCell: FC<TableCellProps> = ({
  className,
  _responsive = false,
  _responsiveHeading = '',
  isNumeric,
  children,
  ...rest
}) => {
  const section = useContext(TableSectionContext);
  useDevWarning(CellOutsideOfSectionWarning, () => section === TableSection.NONE);

  const cellClass = section === TableSection.HEAD ? 'nhsuk-table__header' : 'nhsuk-table__cell';
  const classes = classNames(cellClass, { [`${cellClass}--numeric`]: isNumeric }, className);

  return (
    <>
      {section === TableSection.HEAD ? (
        <th className={classes} scope="col" {...rest}>
          {children}
        </th>
      ) : (
        <td className={classes} role={_responsive ? 'cell' : undefined} {...rest}>
          {_responsive && (
            <span className="nhsuk-table-responsive__heading">{_responsiveHeading}</span>
          )}
          {children}
        </td>
      )}
    </>
  );
};

TableCell.displayName = 'Table.Cell';

export default TableCell;
