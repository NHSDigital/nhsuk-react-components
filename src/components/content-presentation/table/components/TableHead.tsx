import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const TableHead: FC<HTMLProps<HTMLTableSectionElement>> = ({ className, children, ...rest }) => (
  <thead className={classNames('nhsuk-table__head', className)} {...rest}>
    <TableSectionContext.Provider value={TableSection.HEAD}>
      {children}
    </TableSectionContext.Provider>
  </thead>
);

TableHead.displayName = 'Table.Head';

export default TableHead;
