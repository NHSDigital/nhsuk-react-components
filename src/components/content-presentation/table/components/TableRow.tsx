'use client';
import classNames from 'classnames';
import React, { Children, cloneElement, FC, HTMLProps, useContext, useEffect } from 'react';
import TableContext from '../TableContext';
import { getHeadingsFromChildren, isTableCell } from '../TableHelpers';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const TableRow: FC<HTMLProps<HTMLTableRowElement>> = ({ className, children, ...rest }) => {
  const section = useContext(TableSectionContext);
  const { isResponsive, headings, setHeadings } = useContext(TableContext);

  useEffect(() => {
    if (isResponsive && section === TableSection.HEAD) {
      setHeadings(getHeadingsFromChildren(children));
    }
  }, [isResponsive, section, children]);

  if (isResponsive && section === TableSection.BODY) {
    const tableCells = Children.map(children, (child, index) => {
      if (isTableCell(child)) {
        return cloneElement(child, {
          _responsive: isResponsive,
          _responsiveHeading: `${headings[index] || ''} `,
        });
      }
      return child;
    });

    return (
      <tr className={classNames('nhsuk-table__row', className)} {...rest}>
        {tableCells}
      </tr>
    );
  }

  return (
    <tr className={classNames('nhsuk-table__row', className)} {...rest}>
      {children}
    </tr>
  );
};
TableRow.displayName = 'Table.Row';

export default TableRow;
