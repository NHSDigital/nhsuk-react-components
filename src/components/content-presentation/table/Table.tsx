import React, { ComponentProps, FC, HTMLProps, ReactNode, useMemo, useState } from 'react';
import classNames from 'classnames';
import TableBody from './components/TableBody';
import TableCaption from './components/TableCaption';
import TableCell, { TableCellProps } from './components/TableCell';
import TableContainer from './components/TableContainer';
import TableHead from './components/TableHead';
import TableRow from './components/TableRow';
import TablePanel, { TablePanelProps } from './components/TablePanel';
import TableContext, { ITableContext } from './TableContext';

interface TableProps extends HTMLProps<HTMLTableElement> {
  responsive?: boolean;
  caption?: ReactNode;
  captionProps?: ComponentProps<typeof TableCaption>;
}

interface Table extends FC<TableProps> {
  Body: FC<HTMLProps<HTMLTableSectionElement>>;
  Cell: FC<TableCellProps>;
  Container: FC<HTMLProps<HTMLDivElement>>;
  Head: FC<HTMLProps<HTMLTableSectionElement>>;
  Panel: FC<TablePanelProps>;
  Row: FC<HTMLProps<HTMLTableRowElement>>;
}

const Table = ({
  caption,
  captionProps,
  children,
  className,
  responsive = false,
  ...rest
}: TableProps) => {
  const [headings, setHeadings] = useState<string[]>([]);

  const contextValue: ITableContext = useMemo(() => {
    return {
      isResponsive: Boolean(responsive),
      headings,
      setHeadings,
    };
  }, [responsive, headings, setHeadings]);

  return (
    <TableContext.Provider value={contextValue}>
      <table
        className={classNames(
          { 'nhsuk-table': !responsive },
          { 'nhsuk-table-responsive': responsive },
          className,
        )}
        {...rest}
      >
        {caption && <TableCaption {...captionProps}>{caption}</TableCaption>}
        {children}
      </table>
    </TableContext.Provider>
  );
};

Table.Body = TableBody;
Table.Cell = TableCell;
Table.Container = TableContainer;
Table.Head = TableHead;
Table.Panel = TablePanel;
Table.Row = TableRow;

export default Table;
