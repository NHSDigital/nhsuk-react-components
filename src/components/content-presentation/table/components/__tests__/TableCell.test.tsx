import React from 'react';
import { render } from '@testing-library/react';

import Table from '../../Table';
import TableBody from '../TableBody';
import TableCell from '../TableCell';
import TableHead from '../TableHead';
import TableRow from '../TableRow';

describe('Table.Cell', () => {
  it('matches snapshot', () => {
    const { container } = render(<TableCell />);

    expect(container).toMatchSnapshot();
  });

  it('prints dev warning when used outside of a head or body', () => {
    jest.spyOn(console, 'warn').mockImplementation();
    render(
      <table>
        <thead>
          <tr>
            <TableCell />
          </tr>
        </thead>
      </table>,
    );

    // eslint-disable-next-line no-console
    expect(console.warn).toHaveBeenCalledTimes(1);
    // eslint-disable-next-line no-console
    expect(console.warn).toHaveBeenLastCalledWith(
      'Table.Cell used outside of a Table.Head or Table.Body component. Unable to determine section type from context.',
    );
  });

  it('returns th element when inside a Table.Head', () => {
    const { container } = render(
      <Table>
        <TableHead>
          <TableRow>
            <TableCell id="test-id" />
          </TableRow>
        </TableHead>
      </Table>,
    );
    const cellWrapper = container.querySelector('th.nhsuk-table__header');

    expect(cellWrapper).toBeTruthy();
  });

  it('returns td element when inside a Table.Body', () => {
    const { container } = render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell id="test-id" />
          </TableRow>
        </TableBody>
      </Table>,
    );
    const cellWrapper = container.querySelector('td.nhsuk-table__cell');

    expect(cellWrapper).toBeTruthy();
  });

  it('adds responsive heading when _responsive=True', () => {
    const { container } = render(
      <Table responsive>
        <TableHead>
          <TableRow>
            <TableCell>TestHeading</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell id="test-id" />
          </TableRow>
        </TableBody>
      </Table>,
    );
    const cellElement = container.querySelector('td');
    const spanWrapper = container.querySelector('td > span');

    expect(cellElement).toBeTruthy();
    expect(spanWrapper?.textContent).toBe('TestHeading ');
  });

  it('adds the numeric class when isNumeric is true', () => {
    const { container } = render(
      <table>
        <tbody>
          <tr>
            <TableCell data-test="cell" isNumeric />
          </tr>
        </tbody>
      </table>,
    );
    const cell = container.querySelector('td[data-test="cell"].nhsuk-table__cell--numeric');

    expect(cell).toBeTruthy();
  });

  it('adds the numeric header class when isNumeric is true', () => {
    const { container } = render(
      <table>
        <TableHead>
          <tr>
            <TableCell data-test="cell" isNumeric />
          </tr>
        </TableHead>
      </table>,
    );
    const cell = container.querySelector('th[data-test="cell"].nhsuk-table__header--numeric');

    expect(cell).toBeTruthy();
  });

  it('does not add the numeric header when isNumeric is false', () => {
    const { container } = render(
      <table>
        <TableHead>
          <tr>
            <TableCell data-test="header" />
          </tr>
        </TableHead>
        <tbody>
          <tr>
            <TableCell data-test="cell" />
          </tr>
        </tbody>
      </table>,
    );
    const header = container.querySelector('th[data-test="header"].nhsuk-table__header--numeric');
    const cell = container.querySelector('td[data-test="cell"].nhsuk-table__header--numeric');

    expect(header).toBeFalsy();
    expect(cell).toBeFalsy();
  });
});
