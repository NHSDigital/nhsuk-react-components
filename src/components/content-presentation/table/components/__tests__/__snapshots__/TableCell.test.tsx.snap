// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.Cell matches snapshot 1`] = `
<div>
  <td
    class="nhsuk-table__cell"
  />
</div>
`;
